In [1]:
##import packages used in this session
import pandas as pd
from random import sample

## Data collection and cleaning

### COVID-19 cases by county, race, and ethnicity
This data source was obtained from the New York Times public GitHub repository. The unit of observation is state; this data frame reports the number of reported COVID-19 cases by race and ethnicity for counties that provided this information to the CDC. The timeframe is from the beginning of the pandemic until the end of May 2020. https://github.com/nytimes/covid-19-data/issues/381

In [2]:
##import NYT data 
county_data=pd.read_csv("Data/data.csv",dtype={'fips': object})

In [3]:
county_data.shape

(974, 23)

There are 974 counties represented in this data. There are approximately 3,007 counties in the United States. 

In [4]:
## column headers
county_data.columns

Index(['fips', 'state', 'county', 'cases', 'white_cases', 'black_cases',
       'hispanic_cases', 'native_cases', 'asian_cases', 'white_rate',
       'black_rate', 'hispanic_rate', 'native_rate', 'asian_rate', 'known_pct',
       'pop_white', 'pop_black', 'pop_hispanic', 'pct_white', 'pct_black',
       'pct_hispanic', 'pct_asian', 'pct_native'],
      dtype='object')

The NYT data set contains pre-calculated COVID-19 case rates amongst different racial and ethnic populations and also contains population data for white, black, and hispanic populations in each of the 974 counties.Note the population data for asian and native american populations is not included. Since the outcome variable for the statistical learning componenet will rely on population data for all minority populations represented in this data (black, hispanic, native american, and asian), only the COVID-19 cases will remain in this data set, and population and COVID-19 rates for each racial/ethnic category in each county will be calculated later on using the most recent population data from the U.S. Census. 

In [5]:
## keep relevant columsn(i.e., county fips code; state; county; total cases; cases by race)
county_data=county_data.filter(items=['fips', 'state', 'county', 'cases', 'white_cases', 'black_cases',
       'hispanic_cases', 'native_cases', 'asian_cases'])

In [6]:
county_data.head(10)

fips    state           county  cases  white_cases  black_cases  \
0  01001  Alabama   Autauga County    119           44           21   
1  01003  Alabama   Baldwin County    201          144            6   
2  01005  Alabama   Barbour County     91            5           32   
3  01007  Alabama      Bibb County     50           17            5   
4  01009  Alabama    Blount County     37           15            0   
5  01011  Alabama   Bullock County    125            0          106   
6  01013  Alabama    Butler County    263           65          166   
7  01015  Alabama   Calhoun County    124           72           23   
8  01017  Alabama  Chambers County    301           81          181   
9  01021  Alabama   Chilton County     65           27            0   

   hispanic_cases  native_cases  asian_cases  
0               0             0            0  
1               0             0            0  
2               0             0            0  
3               0             0            0  
4               0             0            0  
5               0             0            0  
6               0             0            0  
7               0             0            0  
8               0             0            0  
9               0             0            0

For ease of analysis and visualization later on, the state names will be converted to two-letter abbreviations. A publicly available dictionary, created and made available by Roger Allen (https://gist.github.com/rogerallen/1583593), will be used for this conversion.

In [7]:
##use code provided by roger allen
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'American Samoa': 'AS',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'District of Columbia': 'DC',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Guam': 'GU',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Northern Mariana Islands':'MP',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Pennsylvania': 'PA',
    'Puerto Rico': 'PR',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virgin Islands': 'VI',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY'
}

In [8]:
##convert state names to two-letter abbreviations
county_data['state'] = county_data['state'].apply(us_state_abbrev.get)

In [9]:
##view dataframe; confirm changes.
county_data.sample(20)

fips state               county  cases  white_cases  black_cases  \
404  25017    MA     Middlesex County  21714         8884         2296   
7    01015    AL       Calhoun County    124           72           23   
389  23001    ME  Androscoggin County    219           67          107   
619  36065    NY        Oneida County   1108          254           61   
402  25013    MA       Hampden County   6044         2243          420   
444  26125    MI       Oakland County   8224         3352         2546   
323  19139    IA     Muscatine County    498          217            0   
635  36099    NY        Seneca County     34            7            0   
926  53033    WA          King County   7057         2301          635   
588  34031    NJ       Passaic County  14772         2104          875   
257  17073    IL         Henry County     51           33            0   
417  26027    MI          Cass County     59           30            0   
765  39157    OH    Tuscarawas County    242           70            0   
307  19049    IA        Dallas County    624          249           33   
962  55111    WI          Sauk County     66           30            0   
832  45025    SC  Chesterfield County    121            0           59   
891  51089    VA         Henry County     80           11            5   
587  34029    NJ         Ocean County   8359         3969          255   
478  27085    MN        McLeod County     16            5            0   
862  49011    UT         Davis County    406          279            0   

     hispanic_cases  native_cases  asian_cases  
404            3174             6         1001  
7                 0             0            0  
389               0             0            0  
619               5             0            5  
402             926             0           31  
444             115             5          250  
323             180             0            0  
635               0             0            0  
926            1299            29          855  
588            4706             0          261  
257               0             0            0  
417               0             0            0  
765             125             0            0  
307             240             0            5  
962               7             0            0  
832               0             0            0  
891               0             0            0  
587             648             0          103  
478               0             0            0  
862              37             0            5

In [10]:
##remove leading zeros from fips code
county_data['fips'] = [ i.lstrip('0') for i in county_data['fips'] ]

In [11]:
## check dtypes
county_data.dtypes

fips              object
state             object
county            object
cases              int64
white_cases        int64
black_cases        int64
hispanic_cases     int64
native_cases       int64
asian_cases        int64
dtype: object

### Population: https://www.census.gov/newsroom/press-kits/2020/population-estimates-detailed.html
NYT missing population data for Native American and Asian population.

In [12]:
##population data
county_pop=pd.read_csv("Data/cc-est2019-alldata.csv", encoding='ISO-8859-1', dtype={'COUNTY': object,'STATE': object})

In [13]:
## based on the code book,  2019 is coded as 12 under column YEAR
county_pop=county_pop.loc[county_pop['YEAR'] == 12]

In [14]:
## population estimate for all ages is coded as 0 under column AGEGRP
county_pop=county_pop.loc[county_pop['AGEGRP'] == 0]

In [15]:
## keep only relevant columns
county_pop=county_pop.filter(items=['STATE','COUNTY', 'STNAME', 'CTYNAME','TOT_POP','H_MALE', 'H_FEMALE',
                                    'WA_MALE', 'WA_FEMALE', 'BA_MALE', 'BA_FEMALE', 'IA_MALE', 'IA_FEMALE', 
                                    'AA_MALE', 'AA_FEMALE', 'NA_MALE', 'NA_FEMALE','TOM_MALE', 'TOM_FEMALE' ])

In [16]:
county_pop

STATE COUNTY   STNAME            CTYNAME  TOT_POP  H_MALE  H_FEMALE  \
209       01    001  Alabama     Autauga County    55869     884       787   
437       01    003  Alabama     Baldwin County   223234    5545      4989   
665       01    005  Alabama     Barbour County    24686     629       488   
893       01    007  Alabama        Bibb County    22394     343       280   
1121      01    009  Alabama      Blount County    57826    2950      2632   
...      ...    ...      ...                ...      ...     ...       ...   
715445    56    037  Wyoming  Sweetwater County    42343    3551      3221   
715673    56    039  Wyoming       Teton County    23464    1884      1670   
715901    56    041  Wyoming       Uinta County    20226     927       944   
716129    56    043  Wyoming    Washakie County     7805     565       543   
716357    56    045  Wyoming      Weston County     6927     155       130   

        WA_MALE  WA_FEMALE  BA_MALE  BA_FEMALE  IA_MALE  IA_FEMALE  AA_MALE  \
209       20878      21729     5237       6000      121        145      286   
437       94810     100388     9486      10107      903        839      932   
665        6389       5745     6311       5595      103         67       55   
893        8766       8425     2941       1822       53         50       23   
1121      27258      28154      516        462      192        178       85   
...         ...        ...      ...        ...      ...        ...      ...   
715445    20446      19252      347        251      324        302      203   
715673    11567      10718      101         71      106        102      143   
715901     9753       9524       77         75      135        157       38   
716129     3759       3618       25         19       63         75       25   
716357     3392       3062       32         16       78         53       34   

        AA_FEMALE  NA_MALE  NA_FEMALE  TOM_MALE  TOM_FEMALE  
209           370       32         26       538         507  
437          1448       74         80      2042        2125  
665            61       34         18       172         136  
893            25       22          4       124         139  
1121          100       42         25       379         435  
...           ...      ...        ...       ...         ...  
715445        250       35         34       453         446  
715673        252       20         13       205         166  
715901         62       17         14       204         170  
716129         39        4          2        87          89  
716357         83        2          0        86          89  

[3142 rows x 19 columns]

In [17]:
##combine state and county fips to make consistent with NYT dataframe
county_pop['fips'] = county_pop[['STATE', 'COUNTY']].apply(lambda x: ''.join(x), axis=1)
##remove leading zeros from fips codes
county_pop['fips'] = [ i.lstrip('0') for i in county_pop['fips'] ]

In [18]:
## convert state names to two-letter abbreviations
county_pop['STNAME'] = county_pop['STNAME'].apply(us_state_abbrev.get)

In [19]:
##sum male and female population to get total population for all races and ethnicities 
county_pop['WH']=county_pop['WA_MALE']+county_pop['WA_FEMALE']
county_pop['BA']=county_pop['BA_MALE']+county_pop['BA_FEMALE']
county_pop['AA']=county_pop['AA_MALE']+county_pop['AA_FEMALE']
county_pop['IA']=county_pop['IA_MALE']+county_pop['IA_FEMALE']
county_pop['HI']=county_pop['H_MALE']+county_pop['H_FEMALE']

In [20]:
##keep relevant columns for merging
county_pop=county_pop.filter(items=['fips','TOT_POP','WH','BA','AA','IA','HI'])

In [21]:
##view data frame.
county_pop

fips  TOT_POP      WH     BA    AA    IA     HI
209      1001    55869   42607  11237   656   266   1671
437      1003   223234  195198  19593  2380  1742  10534
665      1005    24686   12134  11906   116   170   1117
893      1007    22394   17191   4763    48   103    623
1121     1009    57826   55412    978   185   370   5582
...       ...      ...     ...    ...   ...   ...    ...
715445  56037    42343   39698    598   453   626   6772
715673  56039    23464   22285    172   395   208   3554
715901  56041    20226   19277    152   100   292   1871
716129  56043     7805    7377     44    64   138   1108
716357  56045     6927    6454     48   117   131    285

[3142 rows x 7 columns]

### Economic indicators

#### County GDP
County level GDP data is obtained from the Bureau of Economic Analysis at the U.S. Department of Commerce. The metric selected is GDP by county and metropolitan area across all industries. This data was last updated on December 12, 2019.
https://apps.bea.gov/iTable/iTable.cfm?reqid=70&step=1&isuri=1&acrdn=5#reqid=70&step=1&isuri=1&acrdn=5

In [22]:
gdp=pd.read_csv("Data/GDP_county - GDP_county.csv")

In [23]:
gdp.tail(11)

GeoFips     GeoName    2018
3115                                              56045  Weston, WY  318545
3116                                Legend / Footnotes:         NaN     NaN
3117  1/ Gross Domestic Product (GDP) is in thousand...         NaN     NaN
3118  * Broomfield County, CO, was created from part...         NaN     NaN
3119  * Estimates from 2008 forward separate Skagway...         NaN     NaN
3120  * Virginia combination areas consist of one or...         NaN     NaN
3121  * Shannon County, SD was renamed to Oglala Lak...         NaN     NaN
3122  * Kalawao County, Hawaii is combined with Maui...         NaN     NaN
3123  Metropolitan Areas are defined (geographically...         NaN     NaN
3124                                (NA) Not available.         NaN     NaN
3125  Last updated: December 12, 2019 -- new statist...         NaN     NaN

Tail of the data frame includes irrelevant rows that needs to be removed before merging this dataframe with the county level COVID-19 cases dataframe.

In [24]:
##remove irrelevant rows
gdp.drop(gdp.tail(10).index, inplace = True) 

In [25]:
## confirm removal
gdp.tail(11)

GeoFips         GeoName     2018
3105   56025     Natrona, WY  5672135
3106   56027    Niobrara, WY   130556
3107   56029        Park, WY  1460391
3108   56031      Platte, WY   577915
3109   56033    Sheridan, WY  1397518
3110   56035    Sublette, WY  1245979
3111   56037  Sweetwater, WY  3880016
3112   56039       Teton, WY  2505534
3113   56041       Uinta, WY   927537
3114   56043    Washakie, WY   379984
3115   56045      Weston, WY   318545

The column headers need to be renamed to maintain consistency across all data frames that will be merged later on. 

In [26]:
##rename column headers
gdp = gdp.rename(columns={'GeoFips':'fips','2018': 'GDP', 'GeoName':'county'})    

In [27]:
##keep only relevant columns
gdp=gdp.filter(items=['fips','GDP','county'])

In [28]:
gdp

fips      GDP          county
0      1001  1690937     Autauga, AL
1      1003  6606080     Baldwin, AL
2      1005   851956     Barbour, AL
3      1007   424510        Bibb, AL
4      1009   942904      Blount, AL
...     ...      ...             ...
3111  56037  3880016  Sweetwater, WY
3112  56039  2505534       Teton, WY
3113  56041   927537       Uinta, WY
3114  56043   379984    Washakie, WY
3115  56045   318545      Weston, WY

[3116 rows x 3 columns]

Unlike the NYT data set, the county and state names are in a single column. This will be remedied by creating a separate column for states. The county column will have the word 'county' added to all observations to maintain consistency with the NYT data set.

In [29]:
## create a separate state column
gdp['state']=gdp['county'].str.split(",").str[1]

In [30]:
##retain only county names in the county column
gdp['county']= gdp['county'].str.split(",").str[0].astype(str)

Unlike the NYT dataframe, the county column in this dataframe does not have the word county. To remedy this, first check if any of the observations in the NYT dataframe do not have the word county. 

In [31]:
##query counties in NYT dataframe without the word county
county_exclude=county_data[~county_data['county'].str.contains(r'County')]
county_exclude=county_exclude['county']
county_exclude

56           Anchorage Municipality
57     Fairbanks North Star Borough
58          Juneau City and Borough
59        Ketchikan Gateway Borough
916                      Salem city
Name: county, dtype: object

In [32]:
## add 'county' to  observations in the county column, excluding the ones queried above
gdp['county'] = gdp['county']+' County'  

In [33]:
##keep only relevant columns for merging
gdp=gdp.filter(items=['fips','GDP'])

In [34]:
## view data frame; confirm changes
gdp

fips      GDP
0      1001  1690937
1      1003  6606080
2      1005   851956
3      1007   424510
4      1009   942904
...     ...      ...
3111  56037  3880016
3112  56039  2505534
3113  56041   927537
3114  56043   379984
3115  56045   318545

[3116 rows x 2 columns]

#### County poverty and median household income
Poverty (poverty estimate for all ages and percent in poverty for all ages) and median household income by county was obtained from  U.S. Census Bureau, Small Area Income and Poverty Estimates (SAIPE) Program (i.e.,2018 Poverty and Median Household Income Estimates). This data was last revised on December 2019.
https://www.census.gov/data/datasets/2018/demo/saipe/2018-state-and-county.html

In [35]:
##import data;read FIPS code columns as dtype object to maintain traiiling zeroes.
pov_inc=pd.read_csv("Data/est18all.xlsx - est18ALL.csv",skiprows=1, dtype={'County FIPS Code': object,'State FIPS Code': object})

In [36]:
##examine shape of data
pov_inc.shape

(3194, 31)

In [37]:
##view random sample of dataframe to understand the distribution of the data
pov_inc.sample(10)

State FIPS Code County FIPS Code Postal Code                Name  \
2800              48              465          TX    Val Verde County   
2415              46              023          SD  Charles Mix County   
2996              51              800          VA        Suffolk city   
1387              27              097          MN     Morrison County   
1730              31              095          NE    Jefferson County   
949               20              087          KS    Jefferson County   
658               17              097          IL         Lake County   
2645              48              155          TX        Foard County   
2958              51              191          VA   Washington County   
2167              39              175          OH      Wyandot County   

     Poverty Estimate, All Ages 90% CI Lower Bound 90% CI Upper Bound  \
2800                      8,950              6,886             11,014   
2415                      1,857              1,459              2,255   
2996                      9,846              7,912             11,780   
1387                      3,307              2,673              3,941   
1730                        821                647                995   
949                       1,411              1,083              1,739   
658                      54,353             47,531             61,175   
2645                        210                159                261   
2958                      8,022              6,691              9,353   
2167                      1,723              1,355              2,091   

     Poverty Percent, All Ages 90% CI Lower Bound.1 90% CI Upper Bound.1  ...  \
2800                      18.9                 14.5                 23.3  ...   
2415                      20.9                 16.4                 25.4  ...   
2996                      10.9                  8.8                 13.0  ...   
1387                      10.1                  8.2                 12.0  ...   
1730                      11.8                  9.3                 14.3  ...   
949                        7.6                  5.8                  9.4  ...   
658                        8.0                  7.0                  9.0  ...   
2645                      18.0                 13.7                 22.3  ...   
2958                      15.2                 12.7                 17.7  ...   
2167                       8.0                  6.3                  9.7  ...   

     90% CI Upper Bound.5 Median Household Income 90% CI Lower Bound.6  \
2800                 39.0                  44,276               39,819   
2415                 36.1                  47,412               43,903   
2996                 20.1                  69,753               65,114   
1387                 14.6                  57,471               54,738   
1730                 18.1                  47,075               42,209   
949                  12.2                  69,246               63,980   
658                  11.0                  87,336               85,186   
2645                 35.4                  37,648               33,934   
2958                 22.2                  45,510               41,540   
2167                 12.1                  55,995               50,857   

     90% CI Upper Bound.6 Poverty Estimate, Age 0-4 90% CI Lower Bound.7  \
2800               48,733                         .                    .   
2415               50,921                         .                    .   
2996               74,392                         .                    .   
1387               60,204                         .                    .   
1730               51,941                         .                    .   
949                74,512                         .                    .   
658                89,486                         .                    .   
2645               41,362                         .                    .   
2958  

In [38]:
## examine column headers
pov_inc.columns

Index(['State FIPS Code', 'County FIPS Code', 'Postal Code', 'Name',
       'Poverty Estimate, All Ages', '90% CI Lower Bound',
       '90% CI Upper Bound', 'Poverty Percent, All Ages',
       '90% CI Lower Bound.1', '90% CI Upper Bound.1',
       'Poverty Estimate, Age 0-17', '90% CI Lower Bound.2',
       '90% CI Upper Bound.2', 'Poverty Percent, Age 0-17',
       '90% CI Lower Bound.3', '90% CI Upper Bound.3',
       'Poverty Estimate, Age 5-17 in Families', '90% CI Lower Bound.4',
       '90% CI Upper Bound.4', 'Poverty Percent, Age 5-17 in Families',
       '90% CI Lower Bound.5', '90% CI Upper Bound.5',
       'Median Household Income', '90% CI Lower Bound.6',
       '90% CI Upper Bound.6', 'Poverty Estimate, Age 0-4',
       '90% CI Lower Bound.7', '90% CI Upper Bound.7',
       'Poverty Percent, Age 0-4', '90% CI Lower Bound.8',
       '90% CI Upper Bound.8'],
      dtype='object')

For the purpose of this analysis, only the 'Poverty Percent,All ages' and the 'Median Household Income' variables will be used.

In [39]:
##keep only relevant columns
pov_inc=pov_inc.filter(items=['State FIPS Code','Name','County FIPS Code','Poverty Percent, All Ages','Median Household Income'])

The format of the county fips code in this data is shortened when compared to the above data frame. The state fips code and county fips code will be combined below to create a consistent fips code. 

In [40]:
##join county and state fips codes
pov_inc['fips'] = pov_inc[['State FIPS Code', 'County FIPS Code']].apply(lambda x: ''.join(x), axis=1)
##remove leading zeros from fips codes
pov_inc['fips'] = [ i.lstrip('0') for i in pov_inc['fips'] ]

In [41]:
##view dataframe to confirm creation of new column
pov_inc.sample(10)

State FIPS Code               Name County FIPS Code  \
201               06      Fresno County              019   
55                01     Pickens County              107   
3019              53      Kitsap County              035   
2565              47  Williamson County              187   
2802              48    Victoria County              469   
769               18       Noble County              113   
1872              36    Columbia County              021   
2391              45      Marion County              067   
2818              48      Yoakum County              501   
1764              31     Sherman County              163   

     Poverty Percent, All Ages Median Household Income   fips  
201                       21.3                  52,318   6019  
55                        23.1                  37,586   1107  
3019                       9.1                  76,530  53035  
2565                       3.8                 115,930  47187  
2802                      15.4                  51,646  48469  
769                        9.1                  61,341  18113  
1872                      11.7                  64,916  36021  
2391                      25.5                  34,365  45067  
2818                      12.1                  61,560  48501  
1764                      14.0                  47,499  31163

Since this data frame does not have state names included, the state fips code column will be converted to 
corresponding two-letter state name. To do this, a dictionary of state fips code and state names will be generated from the dataframe above.

In [42]:
##isolate state fips code and name column for states only into a separate dataframe
fips_conv=pov_inc.loc[pov_inc['County FIPS Code'] == '000']##note:county fips code =='000' for states
fips_conv=fips_conv.filter(items=['State FIPS Code','Name'])##keep relevant columns
fips_conv

State FIPS Code                  Name
0                 00         United States
1                 01               Alabama
69                02                Alaska
99                04               Arizona
115               05              Arkansas
191               06            California
250               08              Colorado
315               09           Connecticut
324               10              Delaware
328               11  District of Columbia
330               12               Florida
398               13               Georgia
558               15                Hawaii
564               16                 Idaho
609               17              Illinois
712               18               Indiana
805               19                  Iowa
905               20                Kansas
1011              21              Kentucky
1132              22             Louisiana
1197              23                 Maine
1214              24              Maryland
1239              25         Massachusetts
1254              26              Michigan
1338              27             Minnesota
1426              28           Mississippi
1509              29              Missouri
1625              30               Montana
1682              31              Nebraska
1776              32                Nevada
1794              33         New Hampshire
1805              34            New Jersey
1827              35            New Mexico
1861              36              New York
1924              37        North Carolina
2025              38          North Dakota
2079              39                  Ohio
2168              40              Oklahoma
2246              41                Oregon
2283              42          Pennsylvania
2351              44          Rhode Island
2357              45        South Carolina
2404              46          South Dakota
2471              47             Tennessee
2567              48                 Texas
2822              49                  Utah
2852              50               Vermont
2867              51              Virginia
3001              53            Washington
3041              54         West Virginia
3097              55             Wisconsin
3170              56               Wyoming

In [43]:
##reshape data frame from long to wide 
fips_conv = fips_conv.set_index("State FIPS Code").T

In [44]:
##view reshaped dataframe
fips_conv

State FIPS Code             00       01      02       04        05  \
Name             United States  Alabama  Alaska  Arizona  Arkansas   

State FIPS Code          06        08           09        10  \
Name             California  Colorado  Connecticut  Delaware   

State FIPS Code                    11  ...            46         47     48  \
Name             District of Columbia  ...  South Dakota  Tennessee  Texas   

State FIPS Code    49       50        51          53             54  \
Name             Utah  Vermont  Virginia  Washington  West Virginia   

State FIPS Code         55       56  
Name             Wisconsin  Wyoming  

[1 rows x 52 columns]

In [45]:
##convert to list dictionary
fips_conv = fips_conv.to_dict('r')

In [46]:
## view list item
fips_conv[0]

{'00': 'United States',
 '01': 'Alabama',
 '02': 'Alaska',
 '04': 'Arizona',
 '05': 'Arkansas',
 '06': 'California',
 '08': 'Colorado',
 '09': 'Connecticut',
 '10': 'Delaware',
 '11': 'District of Columbia',
 '12': 'Florida',
 '13': 'Georgia',
 '15': 'Hawaii',
 '16': 'Idaho',
 '17': 'Illinois',
 '18': 'Indiana',
 '19': 'Iowa',
 '20': 'Kansas',
 '21': 'Kentucky',
 '22': 'Louisiana',
 '23': 'Maine',
 '24': 'Maryland',
 '25': 'Massachusetts',
 '26': 'Michigan',
 '27': 'Minnesota',
 '28': 'Mississippi',
 '29': 'Missouri',
 '30': 'Montana',
 '31': 'Nebraska',
 '32': 'Nevada',
 '33': 'New Hampshire',
 '34': 'New Jersey',
 '35': 'New Mexico',
 '36': 'New York',
 '37': 'North Carolina',
 '38': 'North Dakota',
 '39': 'Ohio',
 '40': 'Oklahoma',
 '41': 'Oregon',
 '42': 'Pennsylvania',
 '44': 'Rhode Island',
 '45': 'South Carolina',
 '46': 'South Dakota',
 '47': 'Tennessee',
 '48': 'Texas',
 '49': 'Utah',
 '50': 'Vermont',
 '51': 'Virginia',
 '53': 'Washington',
 '54': 'West Virginia',
 '55': 'Wis

In [47]:
##convert list to dictionary
fips_conv=dict(fips_conv[0])

In [48]:
##convert fips code to full state name using dictionary
pov_inc['State FIPS Code'] = pov_inc['State FIPS Code'].apply(fips_conv.get)
##view dataframe; confirm change
pov_inc

State FIPS Code               Name County FIPS Code  \
0      United States      United States              000   
1            Alabama            Alabama              000   
2            Alabama     Autauga County              001   
3            Alabama     Baldwin County              003   
4            Alabama     Barbour County              005   
...              ...                ...              ...   
3189         Wyoming  Sweetwater County              037   
3190         Wyoming       Teton County              039   
3191         Wyoming       Uinta County              041   
3192         Wyoming    Washakie County              043   
3193         Wyoming      Weston County              045   

     Poverty Percent, All Ages Median Household Income   fips  
0                         13.1                  61,937         
1                         16.8                  49,881   1000  
2                         13.8                  59,338   1001  
3                          9.8                  57,588   1003  
4                         30.9                  34,382   1005  
...                        ...                     ...    ...  
3189                       8.4                  73,315  56037  
3190                       6.3                  99,087  56039  
3191                      10.0                  63,401  56041  
3192                      11.9                  55,190  56043  
3193                      11.8                  54,319  56045  

[3194 rows x 6 columns]

In [49]:
## convert state name to two-letters; use dictionary from used for NYT dataframe
pov_inc['State FIPS Code'] = pov_inc['State FIPS Code'].apply(us_state_abbrev.get)
##view dataframe; confirm change
pov_inc

State FIPS Code               Name County FIPS Code  \
0               None      United States              000   
1                 AL            Alabama              000   
2                 AL     Autauga County              001   
3                 AL     Baldwin County              003   
4                 AL     Barbour County              005   
...              ...                ...              ...   
3189              WY  Sweetwater County              037   
3190              WY       Teton County              039   
3191              WY       Uinta County              041   
3192              WY    Washakie County              043   
3193              WY      Weston County              045   

     Poverty Percent, All Ages Median Household Income   fips  
0                         13.1                  61,937         
1                         16.8                  49,881   1000  
2                         13.8                  59,338   1001  
3                          9.8                  57,588   1003  
4                         30.9                  34,382   1005  
...                        ...                     ...    ...  
3189                       8.4                  73,315  56037  
3190                       6.3                  99,087  56039  
3191                      10.0                  63,401  56041  
3192                      11.9                  55,190  56043  
3193                      11.8                  54,319  56045  

[3194 rows x 6 columns]

In [50]:
## keep only relevant columns
pov_inc=pov_inc.filter(items=['fips','Poverty Percent, All Ages','Median Household Income'])
 

In [51]:
## rename column headers for consistency across data frames
pov_inc = pov_inc.rename(columns={'Poverty Percent, All Ages': 'pov_perc',
                                  'Median Household Income': 'Med_inc'})   

In [52]:
##view data frame
pov_inc

fips pov_perc Med_inc
0               13.1  61,937
1      1000     16.8  49,881
2      1001     13.8  59,338
3      1003      9.8  57,588
4      1005     30.9  34,382
...     ...      ...     ...
3189  56037      8.4  73,315
3190  56039      6.3  99,087
3191  56041     10.0  63,401
3192  56043     11.9  55,190
3193  56045     11.8  54,319

[3194 rows x 3 columns]

#### Unemployment Rate
County level unemployment rate is obtained from the U.S. Bureau of Labor Statistics. The data table below represents 2019 annual average labor force data by county.  
https://www.bls.gov/lau/tables.htm

In [53]:
##import data; read FIPS code columns as dtype object to maintain traiiling zeroes.
unemp=pd.read_csv("Data/laucnty19.xlsx - laucnty19-2.csv", dtype={'County FIPS Code': object,'State FIPSCode': object})

In [54]:
unemp

Code State FIPSCode County FIPS Code  \
0     CN0100100000000             01              001   
1     CN0100300000000             01              003   
2     CN0100500000000             01              005   
3     CN0100700000000             01              007   
4     CN0100900000000             01              009   
...               ...            ...              ...   
3214  CN7214500000000             72              145   
3215  CN7214700000000             72              147   
3216  CN7214900000000             72              149   
3217  CN7215100000000             72              151   
3218  CN7215300000000             72              153   

     County Name/State Abbreviation  Year  Labor Force     Employed  \
0                Autauga County, AL  2019  26,172       25,458        
1                Baldwin County, AL  2019  97,328       94,675        
2                Barbour County, AL  2019   8,537        8,213        
3                   Bibb County, AL  2019   8,685        8,419        
4                 Blount County, AL  2019  25,331       24,655        
...                             ...   ...          ...          ...   
3214        Vega Baja Municipio, PR  2019  13,037       11,791        
3215          Vieques Municipio, PR  2019   2,585        2,406        
3216         Villalba Municipio, PR  2019   7,406        6,231        
3217          Yabucoa Municipio, PR  2019   8,691        7,552        
3218            Yauco Municipio, PR  2019   9,759        8,331        

      Unemployed  Unemployment Rate  
0       714                     2.7  
1     2,653                     2.7  
2       324                     3.8  
3       266                     3.1  
4       676                     2.7  
...          ...                ...  
3214  1,246                     9.6  
3215    179                     6.9  
3216  1,175                    15.9  
3217  1,139                    13.1  
3218  1,428                    14.6  

[3219 rows x 9 columns]

In [55]:
##keep only relevant columns
unemp=unemp.filter(items=['State FIPSCode','County FIPS Code','Unemployment Rate'])

In [56]:
unemp

State FIPSCode County FIPS Code  Unemployment Rate
0                01              001                2.7
1                01              003                2.7
2                01              005                3.8
3                01              007                3.1
4                01              009                2.7
...             ...              ...                ...
3214             72              145                9.6
3215             72              147                6.9
3216             72              149               15.9
3217             72              151               13.1
3218             72              153               14.6

[3219 rows x 3 columns]

Similar to the previous data frames, the county and state names are combined into a single column and the state and fips code are in a separate column. This will be remedied using the same tools as above

In [57]:
##join county and state fips codes
unemp['fips'] = unemp[['State FIPSCode', 'County FIPS Code']].apply(lambda x: ''.join(x), axis=1)
##remove leading zeros from fips codes
unemp['fips'] = [ i.lstrip('0') for i in unemp['fips'] ]

In [58]:
##keep only relevant columns for merge
unemp=unemp.filter(items=['fips','Unemployment Rate'])

In [59]:
##view dataframe
unemp 

fips  Unemployment Rate
0      1001                2.7
1      1003                2.7
2      1005                3.8
3      1007                3.1
4      1009                2.7
...     ...                ...
3214  72145                9.6
3215  72147                6.9
3216  72149               15.9
3217  72151               13.1
3218  72153               14.6

[3219 rows x 2 columns]

### County characteristics: Social, phyisical, environmental, and clinical factors
The county health rankings and roadmaps program is a collaboration between the Robert Wood Johnson Foundation and the University of Wisconsin Population Health Institute. This program collects data on a range of county-level metrics such as the number of uninsured adults in a county; average traffic volume in a county, and other metrics that are used to assess the overall health of a county. Since this analysis is aimed at exploring the county-level factors that contribute to racial and ethnic disparities in COVID-19 cases, the variables in this vast dataset  can serve as vital features for the statistical learning portion. The variables selected serve as proxies for social-determinants of health that are the root-causes of health disparities.
The variables are: 
Food environment index
Access to exercise opportunities
Income inequality
Violent crime rate


https://www.countyhealthrankings.org/explore-health-rankings/rankings-data-documentation

In [60]:
##import county ranking data; read FIPS code as object to maintain leading and trailing zeroes
county_rank=pd.read_csv("Data/analytic_data2020_0.csv", dtype={'5-digit FIPS Code': object})

/Users/mkbelay/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (0,1,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,2

In [61]:
##shape of data
county_rank.shape

(3195, 786)

In [62]:
county_rank=county_rank.filter(items=["5-digit FIPS Code","Traffic volume raw value",
                                      "Severe housing cost burden raw value","Homeownership raw value",
                                      "Residential segregation - non-White/White raw value",
                                      "Food environment index raw value",
                                      "Ratio of population to primary care physicians.",
                                      "Drinking water violations raw value",
                                      "Air pollution - particulate matter raw value","Uninsured adults raw value",
                                     "Access to exercise opportunities raw value","Income inequality raw value",
                                      "Violent crime raw value"])

In [63]:
county_rank

5-digit FIPS Code Traffic volume raw value  \
0             fipscode            v156_rawvalue   
1                00000                      NaN   
2                01000             166.00847228   
3                01001             88.457040416   
4                01003             86.997429882   
...                ...                      ...   
3190             56037                  154.755   
3191             56039                  135.189   
3192             56041                  96.1916   
3193             56043                  82.8222   
3194             56045                  64.3043   

     Severe housing cost burden raw value Homeownership raw value  \
0                           v154_rawvalue           v153_rawvalue   
1                            0.1484650658            0.6384759732   
2                            0.1265381674            0.6855046233   
3                            0.1340348117            0.7489462467   
4                            0.1172580795            0.7361934319   
...                                   ...                     ...   
3190                             0.083856                0.753702   
3191                             0.107691                0.577309   
3192                            0.0770767                0.752683   
3193                            0.0681678                0.768264   
3194                             0.110503                0.776943   

     Residential segregation - non-White/White raw value  \
0                                         v142_rawvalue    
1                                           46.77346382    
2                                          50.777775905    
3                                          23.628395199    
4                                          31.825343231    
...                                                 ...    
3190                                            25.3527    
3191                                            29.0447    
3192                                            11.5779    
3193                                            10.8665    
3194                                            9.75976    

     Food environment index raw value  \
0                       v133_rawvalue   
1                                 7.6   
2                                 5.8   
3                                 7.2   
4                                   8   
...                               ...   
3190                              7.7   
3191                              8.2   
3192                              7.4   
3193                              8.3   
3194                              7.9   

     Ratio of population to primary care physicians.  \
0                                  v004_other_data_1   
1                                       1325.0905296   
2                                       1542.6414557   
3                                            2220.16   
4                                       1371.7935484   
...                                              ...   
3190                                         2720.88   
3191                                         861.667   
3192                                         2277.22   
3193                                            2016   
3194                                            2309   

     Drinking water violations raw value  \
0                          v124_rawvalue   
1                                    NaN   
2                           0.1492537313   
3                                      0   
4                                      0   
...                                  ...   
3190                                   1   
3191                                   1   
3192                                   1   
3193                                   0   
3194                                   0   

     Air pollution - particulate matter raw value Uninsured adults raw value  \
0                                   v125_rawvalue              v003

In [64]:
## remove the first row
county_rank=county_rank.iloc[1:]

In [65]:
##reset index
county_rank=county_rank.reset_index(drop=True)

In [66]:
##rename column headers
county_rank=county_rank.rename(columns={'5-digit FIPS Code':'fips',
                                       'Traffic volume raw value':'traffic_vol',
                                        'Severe housing cost burden raw value':'house_burden',
                                        'Homeownership raw value':'ownership',
                                        'Residential segregation - non-White/White raw value':'residential_seg',
                                       'Food environment index raw value':'food_index',
                                        'Ratio of population to primary care physicians.':'pop_to_phys',
                                       'Drinking water violations raw value':'water_violation',
                                        'Air pollution - particulate matter raw value':'air_pollution',
                                       'Access to exercise opportunities raw value':'exercise_opp',
                                        'Income inequality raw value':'income_ineq',
                                        'Violent crime raw value':'crime_violent'})

In [67]:
##removing leading zero from fips column
county_rank['fips'] = [ i.lstrip('0') for i in county_rank['fips'] ]

In [68]:
##view dataframe
county_rank

fips   traffic_vol  house_burden     ownership residential_seg  \
0                     NaN  0.1484650658  0.6384759732     46.77346382   
1      1000  166.00847228  0.1265381674  0.6855046233    50.777775905   
2      1001  88.457040416  0.1340348117  0.7489462467    23.628395199   
3      1003  86.997429882  0.1172580795  0.7361934319    31.825343231   
4      1005  102.29176221  0.1405954631  0.6139777923    23.449712509   
...     ...           ...           ...           ...             ...   
3189  56037       154.755      0.083856      0.753702         25.3527   
3190  56039       135.189      0.107691      0.577309         29.0447   
3191  56041       96.1916     0.0770767      0.752683         11.5779   
3192  56043       82.8222     0.0681678      0.768264         10.8665   
3193  56045       64.3043      0.110503      0.776943         9.75976   

     food_index   pop_to_phys water_violation air_pollution  \
0           7.6  1325.0905296             NaN           8.6   
1           5.8  1542.6414557    0.1492537313            11   
2           7.2       2220.16               0          11.7   
3             8  1371.7935484               0          10.3   
4           5.6       3158.75               0          11.5   
...         ...           ...             ...           ...   
3189        7.7       2720.88               1           5.1   
3190        8.2       861.667               1           4.9   
3191        7.4       2277.22               1           5.9   
3192        8.3          2016               0           4.8   
3193        7.9          2309               0           4.1   

     Uninsured adults raw value  exercise_opp   income_ineq crime_violent  
0                  0.1224754716  0.8416869169  4.9200178008  386.46489648  
1                  0.1408970991  0.6111228737  5.2611357399  479.91918191  
2                  0.1114898915  0.6913012406  5.2345972691  272.28222006  
3                  0.1434852477  0.7371354895  4.4177666786  203.66039629  
4                  0.1608583414  0.5316676986  5.6814100186  414.27786068  
...                         ...           ...           ...           ...  
3189                   0.169401      0.896863       3.97592       300.494  
3190                   0.175177      0.997182       3.77527           NaN  
3191                   0.165985      0.840184       4.35006       71.0065  
3192                   0.193353      0.831829       3.32746       78.2661  
3193                   0.151172      0.630272       5.14567       157.129  

[3194 rows x 13 columns]

## Merging
Now that the data frames have been cleaned and column headers (specifically for fips and county) have been standardized, the data frame will be combined into one with the NYT data set serving as the main/basis data frame.

####  NYT + Population (master)

In [69]:
##master dataframe will be called df; start with merging NYT dataframe and population dataframe
df=pd.merge(left=county_data, right=county_pop, how='left',on=["fips"], indicator=True)

In [70]:
##view left-only merges; 
df.loc[df._merge=="left_only",:].drop(columns= "_merge")

Empty DataFrame
Columns: [fips, state, county, cases, white_cases, black_cases, hispanic_cases, native_cases, asian_cases, TOT_POP, WH, BA, AA, IA, HI]
Index: []

All observations in the NYT COVID-19 dataframe have been merged with their corresponding population data in the county population dataframe.

In [71]:
##drop _merge column
df=df.drop(columns= "_merge")

In [72]:
##view dataframe
df.sample(10)

fips state             county  cases  white_cases  black_cases  \
455  26161    MI   Washtenaw County   1295          662          391   
530  28071    MS   Lafayette County     91           15           22   
650  37009    NC        Ashe County     15            0            0   
275  17133    IL      Monroe County     79           44            0   
83    5071    AR     Johnson County     31            5            0   
905  51159    VA    Richmond County    251           70           88   
914  51191    VA  Washington County     32           20            0   
354  20119    KS       Meade County     16            0            0   
879  51035    VA     Carroll County     70           38            0   
506  28011    MS     Bolivar County    124            0           75   

     hispanic_cases  native_cases  asian_cases  TOT_POP      WH     BA     AA  \
455               6             0           15   367601  272901  45256  34469   
530               0             0            0    54019   39098  12809   1182   
650               5             0            0    27203   26352    262    138   
275               0             0            0    34637   33834    153    209   
83                0             0            0    26578   24345    583    721   
905              15             0            0     9023    6023   2675     64   
914               0             0            0    53740   51980    816    291   
354               5             0            0     4033    3842     44     36   
879              11             0            0    29791   28964    263     68   
506               0             0            0    30628   10463  19524    346   

       IA     HI  
455  1436  18075  
530   177   1442  
650   114   1390  
275    87    570  
83    328   3791  
905    50    658  
914   128    877  
354    52    763  
879    98   1147  
506    51    684

### Master +GDP

In [73]:
## merge master dataframe from above with GDP data
df = pd.merge(left=df, right=gdp, how='left',on=["fips"], indicator=True)

In [74]:
##view left-only merges; 
df.loc[df._merge=="left_only",:]

fips state                 county  cases  white_cases  black_cases  \
873  51003    VA       Albemarle County    166           76           10   
875  51015    VA         Augusta County     80           53            0   
879  51035    VA         Carroll County     70           38            0   
882  51059    VA         Fairfax County  10209         1308          569   
886  51069    VA       Frederick County    292          121            0   
888  51081    VA     Greensville County     54            0           10   
891  51089    VA           Henry County     80           11            5   
892  51095    VA      James City County    179          113            0   
897  51121    VA      Montgomery County     51           11            0   
901  51143    VA    Pittsylvania County     28            6            0   
903  51149    VA   Prince George County    114            5           23   
904  51153    VA  Prince William County   4981          420          403   
906  51161    VA         Roanoke County     80           34            0   
907  51165    VA      Rockingham County    418          182            0   
909  51175    VA     Southampton County    134           17           37   
910  51177    VA    Spotsylvania County    459           89           10   
914  51191    VA      Washington County     32           20            0   
915  51199    VA            York County     51            7            0   
916  51775    VA             Salem city     20            5            0   

     hispanic_cases  native_cases  asian_cases  TOT_POP      WH      BA  \
873               0             0            0   109330   89388   10600   
875               0             0            0    75558   70102    3518   
879              11             0            0    29791   28964     263   
882            5073             0          722  1147532  742178  121954   
886              47             0            0    89313   80867    4225   
888               0             0            0    11336    4281    6823   
891               0             0            0    50557   37625   11434   
892               0             0            0    76523   61473   10433   
897               0             0            0    98535   85207    4238   
901               0             0            0    60354   45990   12970   
903               0             0            0    38353   23121   12510   
904            1848             6          116   470335  293398  104512   
906               0             0            0    94186   82798    6032   
907             110             0            0    81948   77183    2082   
909               0             0            0    17631   10982    6125   
910             126             0            5   136215  102759   23797   
914               0             0            0    53740   51980     816   
915               0             0            0    68280   51465    9442   
916               0             0            0    25301   22190    1965   

         AA    IA      HI  GDP     _merge  
873    6051   383    6313  NaN  left_only  
875     490   191    2374  NaN  left_only  
879      68    98    1147  NaN  left_only  
882  231085  6108  189515  NaN  left_only  
886    1573   431    8325  NaN  left_only  
888      79    32     316  NaN  left_only  
891     254   180    2926  NaN  left_only  
892    2023   336    4692  NaN  left_only  
897    6245   307    3424  NaN  left_only  
901     303   199    1663  NaN  left_only  
903     837   284    3316  NaN  left_only  
904   44429  5025  115173  NaN  left_only  
906    3282   206    3114  NaN  left_only  
907     800   496    6100  NaN  left_only  
909      90    95     359  NaN  left_only  
910    3859   640   14603  NaN  left_only  
914     291   128     877  NaN  left_only  
915    4147   357    4701  NaN  left_only  
916     475   104     903  NaN  left_only

About 20 counties in VA seem to be missing from the GDP data. Let's query these counties in the GDP using the fips code to see if it is a naming issue or if these observations are not present in the GDP dataframe

In [75]:
## isolate left_only merges
merge_fail_gdp=df.loc[df._merge=="left_only",:]

In [76]:
##turn column of fips to list
merge_fail_gdp=merge_fail_gdp["fips"]. tolist() 
merge_fail_gdp

['51003',
 '51015',
 '51035',
 '51059',
 '51069',
 '51081',
 '51089',
 '51095',
 '51121',
 '51143',
 '51149',
 '51153',
 '51161',
 '51165',
 '51175',
 '51177',
 '51191',
 '51199',
 '51775']

In [77]:
##check if the fips code of observations in the left-only merge are in the GDP data
gdp[gdp['fips'].isin(merge_fail_gdp)]

Empty DataFrame
Columns: [fips, GDP]
Index: []

Looks like these observations are not in the GDP dataframe. Let's do a spot check to make sure

In [78]:
##fips code that did not merge
merge_fail_gdp

['51003',
 '51015',
 '51035',
 '51059',
 '51069',
 '51081',
 '51089',
 '51095',
 '51121',
 '51143',
 '51149',
 '51153',
 '51161',
 '51165',
 '51175',
 '51177',
 '51191',
 '51199',
 '51775']

In [79]:
##create a list to house random sample of merge failed fip codes
merge_fail_samp=sample(merge_fail_gdp,5)

In [80]:
##view random sample of fips codes
merge_fail_samp

['51177', '51175', '51059', '51069', '51089']

In [81]:
##check if random sample of fips codes are in gdp dataframe
gdp.loc[gdp['fips'].isin(merge_fail_samp)]

Empty DataFrame
Columns: [fips, GDP]
Index: []

Since these fips codes are not in the GDP data, the _merge column will be dropped and the resot of the dataframe will be merged

In [82]:
df=df.drop(columns= "_merge")

### Master + Poverty & Median Household Income

In [83]:
## merge master dataframe from above with poverty and median household income data
df = pd.merge(left=df, right=pov_inc, how='left',on=["fips"], indicator=True)

In [84]:
##view left-only merges; 
df.loc[df._merge=="left_only",:]

Empty DataFrame
Columns: [fips, state, county, cases, white_cases, black_cases, hispanic_cases, native_cases, asian_cases, TOT_POP, WH, BA, AA, IA, HI, GDP, pov_perc, Med_inc, _merge]
Index: []

Merge was successful.

In [85]:
##drop _merge column
df=df.drop(columns= "_merge")

### Master + Unemployment Rate

In [86]:
## merge master dataframe from above with Unemployment data
df=pd.merge(left=df, right=unemp, how='left',on=['fips'], indicator=True)

In [87]:
##view left-only merges; 
df.loc[df._merge=="left_only",:]

Empty DataFrame
Columns: [fips, state, county, cases, white_cases, black_cases, hispanic_cases, native_cases, asian_cases, TOT_POP, WH, BA, AA, IA, HI, GDP, pov_perc, Med_inc, Unemployment Rate, _merge]
Index: []

Merge was successful.

In [88]:
##drop _merge column
df=df.drop(columns= "_merge")

### Master + County Rank Metrics

In [89]:
## merge master dataframe from above with county rankings data
df = pd.merge(left=df, right=county_rank, how='left',on=["fips"], indicator=True)

In [90]:
##view left-only merges; 
df.loc[df._merge=="left_only",:]

Empty DataFrame
Columns: [fips, state, county, cases, white_cases, black_cases, hispanic_cases, native_cases, asian_cases, TOT_POP, WH, BA, AA, IA, HI, GDP, pov_perc, Med_inc, Unemployment Rate, traffic_vol, house_burden, ownership, residential_seg, food_index, pop_to_phys, water_violation, air_pollution, Uninsured adults raw value, exercise_opp, income_ineq, crime_violent, _merge]
Index: []

[0 rows x 32 columns]

Merge was successful.

In [91]:
##drop _merge column
df=df.drop(columns= "_merge")

In [92]:
## view final dataframe
df.sample(20)

fips state                county  cases  white_cases  black_cases  \
504  28005    MS          Amite County     40            0            5   
54    1131    AL         Wilcox County     96            0           65   
476  27079    MN       Le Sueur County     32            5            0   
852  45073    SC         Oconee County     29            6            0   
902  51147    VA  Prince Edward County    113            0           28   
730  39043    OH           Erie County    114           43           15   
620  36067    NY       Onondaga County   2277          698          306   
308  19061    IA        Dubuque County    198           90            0   
650  37009    NC           Ashe County     15            0            0   
514  28029    MS         Copiah County    244            5          167   
814  42087    PA        Mifflin County     41           22            0   
170  12105    FL           Polk County    641          285          134   
186  13047    GA        Catoosa County     59           10            0   
362  21001    KY          Adair County     68           46            0   
123   8107    CO          Routt County     45           18            0   
265  17095    IL           Knox County     75           26           27   
720  39011    OH       Auglaize County     28            5            0   
861  49005    UT          Cache County     58           22            0   
167  12099    FL     Palm Beach County   4026         1220          742   
323  19139    IA      Muscatine County    498          217            0   

     hispanic_cases  native_cases  asian_cases  TOT_POP  ...     ownership  \
504               0             0            0    12297  ...       0.85049   
54                0             0            0    10373  ...  0.6495793901   
476               0             0            0    28887  ...      0.810557   
852               0             0            0    79546  ...      0.722264   
902               0             0            0    22802  ...      0.668986   
730               0             0            0    74266  ...      0.690138   
620               0             0           42   460528  ...      0.648343   
308               0             0           33    97311  ...  0.7236107488   
650               5             0            0    27203  ...        0.7518   
514               0             0            0    28065  ...      0.786553   
814               0             0            0    46138  ...      0.703729   
170             102             0            0   724777  ...  0.6852676641   
186               0             0            0    67580  ...   0.750121773   
362               0             0            0    19202  ...  0.7625356125   
123               0             0            0    25638  ...   0.704459799   
265               0             0            0    49699  ...  0.6542006721   
720               0             0            0    45656  ...      0.764995   
861               0             0            0   128289  ...      0.626139   
167            1122             0            0  1496770  ...  0.6861145972   
323             180             0            0    42664  ...  0.7288945331   

     residential_seg  food_index   pop_to_phys  water_violation air_pollution  \
504          14.0656         6.3          4149                0             9   
54      17.867719054         2.5       2679.75                0          10.4   
476          30.5329         9.1       7027.75                0           7.7   
852          35.4305           8       1884.63                0          10.2   
902          11.2454         6.6       1621.64                0             9   
730          56.1769         7.2       1312.58                1          11.1   
620          55.5667         8.1       1060.13                1           8.2   
308     43.885742482         8.1  1260.2727273                0          10.1   
650           47.119         8.1       2246.42                

### Export complete dataframe to csv file
The data collection and high-level cleaning process is complete. The merged dataframe will be exported as a csv file to be used for visualizations and modeling. 

In [93]:
##export as csv file
df.to_csv("Data/COVID_project_PPOL564.csv", index=False)

**Since there will be multiple modeling attempts with different features, a second csv file with 15 additional variables from the county rankings and roadmaps data set will be created.**

In [94]:
##import county ranking data; read FIPS code as object to maintain leading and trailing zeroes
county_rank=pd.read_csv("Data/analytic_data2020_0.csv", dtype={'5-digit FIPS Code': object})

/Users/mkbelay/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (0,1,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,2

In [95]:
## select the additional 15 variables 
county_rank=county_rank.filter(items=["5-digit FIPS Code","High school graduation raw value","Some college raw value",
                                     "Social associations raw value","Percentage of households with high housing costs",
                                     "Percentage of households with overcrowding","Percentage of households with lack of kitchen or plumbing facilities",
                                     "Long commute - driving alone raw value","Life expectancy raw value","Premature age-adjusted mortality raw value",
                                     "Frequent physical distress raw value","Frequent mental distress raw value","Diabetes prevalence raw value","Food insecurity raw value",
                                     "Limited access to healthy foods raw value","% Rural raw value"])

In [96]:
## remove the first row
county_rank=county_rank.iloc[1:]

In [97]:
##reset index
county_rank=county_rank.reset_index(drop=True)

In [98]:
##rename column headers
county_rank=county_rank.rename(columns={'5-digit FIPS Code':'fips',
                                        "High school graduation raw value":"high_sch_grad",
                                        "Some college raw value":"some_college",
                                     "Social associations raw value":"soc_association",
                                        "Percentage of households with high housing costs":"perc_high_cost",
                                     "Percentage of households with overcrowding":"perc_overcrowding",
                                        "Percentage of households with lack of kitchen or plumbing facilities":"perc_plumb",
                                     "Long commute - driving alone raw value":"lng_commute",
                                        "Life expectancy raw value":"life_exp",
                                        "Premature age-adjusted mortality raw value":"pre_mortality",
                                     "Frequent physical distress raw value":"phys_distress",
                                        "Frequent mental distress raw value":"ment_distress",
                                        "Diabetes prevalence raw value":"diab_prevalence",
                                        "Food insecurity raw value":"food_insecure",
                                     "Limited access to healthy foods raw value":"food_access",
                                        "% Rural raw value":"perc_rural"})

In [99]:
##removing leading zero from fips column
county_rank['fips'] = [ i.lstrip('0') for i in county_rank['fips'] ]

### Master + County Rank Metrics (additional 15 variables added)

In [100]:
## merge master dataframe from above with county rankings data
df = pd.merge(left=df, right=county_rank, how='left',on=["fips"], indicator=True)

In [101]:
##view left-only merges; 
df.loc[df._merge=="left_only",:]

Empty DataFrame
Columns: [fips, state, county, cases, white_cases, black_cases, hispanic_cases, native_cases, asian_cases, TOT_POP, WH, BA, AA, IA, HI, GDP, pov_perc, Med_inc, Unemployment Rate, traffic_vol, house_burden, ownership, residential_seg, food_index, pop_to_phys, water_violation, air_pollution, Uninsured adults raw value, exercise_opp, income_ineq, crime_violent, high_sch_grad, some_college, soc_association, perc_high_cost, perc_overcrowding, perc_plumb, lng_commute, life_exp, pre_mortality, phys_distress, ment_distress, diab_prevalence, food_insecure, food_access, perc_rural, _merge]
Index: []

[0 rows x 47 columns]

Merge was successful.

In [102]:
##drop _merge column
df=df.drop(columns= "_merge")

In [103]:
## view final dataframe
df.sample(20)

fips state            county  cases  white_cases  black_cases  \
482  27099    MN      Mower County    231           47           13   
239  16027    ID     Canyon County    264           80            0   
766  39163    OH     Vinton County     10            5            0   
831  45023    SC    Chester County     39           11            0   
218  13213    GA     Murray County     44           10            0   
294  19005    IA  Allamakee County     86           11            0   
731  39045    OH  Fairfield County    193           93           23   
332  19171    IA       Tama County    293          119            0   
267  17099    IL    LaSalle County    118           75            7   
524  28059    MS    Jackson County    255           88          105   
26    1067    AL      Henry County     38            5           10   
33    1085    AL    Lowndes County    144            0          105   
71    4027    AZ       Yuma County    654           44            5   
240  16031    ID     Cassia County     39            5            0   
463  27025    MN    Chisago County     44           11            0   
116   8069    CO    Larimer County    499          204            0   
206  13129    GA     Gordon County     71           31            0   
916  51775    VA        Salem city     20            5            0   
746  39099    OH   Mahoning County    989          734          101   
173  12115    FL   Sarasota County    517          380           19   

     hispanic_cases  native_cases  asian_cases  TOT_POP  ...    perc_plumb  \
482              15             0           60    40062  ...    0.00583847   
239              15             0            0   229849  ...  0.0093812821   
766               0             0            0    13085  ...    0.00890688   
831               0             0            0    32244  ...    0.00652263   
218              10             0            0    40096  ...  0.0027339642   
294              42             0            0    13687  ...  0.0164043082   
731               0             0            0   157574  ...    0.00888647   
332              72             0            0    16854  ...  0.0051967335   
267               0             0            0   108669  ...  0.0066486365   
524               0             0            0   143617  ...    0.00555942   
26                0             0            0    17205  ...  0.0057101025   
33                0             0            0     9726  ...  0.0235849057   
71              477             0            0   213787  ...  0.0083891435   
240               0             0            0    24030  ...  0.0103694102   
463               0             0            0    56579  ...    0.00303183   
116              84             0            0   356899  ...  0.0107031834   
206               0             0            0    57963  ...  0.0085945399   
916               0             0            0    25301  ...    0.00745527   
746               0             0            0   228683  ...     0.0129251   
173              18             0            0   433742  ...  0.0147800085   

     lng_commute      life_exp  pre_mortality  phys_distress ment_distress  \
482          0.2       80.2872        286.856      0.0966533      0.103517   
239        0.366  78.944664264   324.26430559   0.1246872505  0.1245355263   
766        0.504       74.5179        503.484       0.139943      0.151875   
831        0.472       73.5929        576.579       0.132626      0.143447   
218        0.393   75.47106737   494.66820328   0.1190167039  0.1301791808   
294          0.3  79.701410662   294.67406544   0.0968522644   0.112058666   
731        0.476       78.3122        331.615       0.108663      0.124236   
332        0.321  76.902288904   392.33965452   0.0973146501  0.1113799393   
267         0.29  77.118690637   396.07480315   0.1074703244  0.1168412101   
524        0.345       76.3228        434.336       0.129303      0.127795   
26         0.371 

In [104]:
##export second version of df as separate csv file
df.to_csv("Data/COVID_project_PPOL564_v2.csv", index=False)